### DISCUSSION TITLE
[ARCHIVED] Абстракции методологии, их цели и нейминг

### DISCUSSION THREAD

**Discussion by azinit** [feature-sliced/documentation#31] [Archive]
# Дискуссия закрыта, расписывается по #61

---
---
---

Пока что похоже, на такую картину (но надо придти к одному неймингу по итогу):

> См. [что писал про абстракции разного уровня](https://github.com/feature-sliced/wiki/discussions/23#discussioncomment-451269)

### High-level abstractions
> Основные, располагаются на высшем уровне, от наиболее "знающих" к наименее (app => shared)

- `app` - логика инициализации приложения + глобальные стили + провайдеры
- `processes` - надстраничные процессы (авторизация, оформление заказа, ...), контролируют все что ниже
- `pages` - страницы (максимально просты, композируют фичи внутри себя)
- `features` - конкретная функциональность, представляющая собой бизнес-ценность
- `entities` - сущности БЛ, на которых строится логика фич и абстракций повыше (ч-з компоненты, либы и т.д.)
- :fire: `shared` (vs `lib` vs `common`) - общеиспользуемые сущности, решающие чисто **инфраструктурные проблемы**, отстранены от БЛ

### Middle-level abstractions
> Вкладываются в high-level абстракции, чаще всего служат "подпорным материалом" в реализации логики, равнозначны - каждый выполняет свою роль

- :fire: `ui` (или `components`) - UIKit / компоненты *данной* абстракции ("образ" функциональности)
- :fire: `models` (или `store`) - Data-слой абстракции для работы со стейтом / событиями - на что могут опираться более высокоуровневые абстракции ("идея" функциональности")
- `lib`  - ~~структурированный набор хелперов~~ набор функций и инфраструктурных решений для решения конкретной проблемы (например для работы с датами, но зависит от абстракции-родителя)
- `api` - логика работы с API (либо общие, либо конкретные для данной фичи/страницы/...)

## Что хотим?

- [ ] Понять, точно ли мы понимаем одинаково абстракции
   > Есть ли лишние/недостающие, Так ли выстроены связи, Все ли так - как описано выше вообще
- [ ] Определиться с единым неймингом
   > Пока что у нас конфликтуют (где-то взять одно, где-то подобрать общий синоним)
   > - `ui` vs `components` 
   >     1) @karina-drummer , @sergeysova, @binjospookie
   >     2) @martis-git , @niyazm524 
   > - `models` vs `store`
   >     1) @karina-drummer , @sergeysova, @binjospookie
   >     2) @martis-git , @niyazm524 
   > - `shared` vs `lib` vs `common` vs `ui/lib(s)` (оба на верхнем уровне)
   >     1) @sergeysova , @martis-git , @niyazm524 , @binjospookie
   >     2) ???
   >     3) @karina-drummer
   >     4) @ilyaagarkov

---

<details>
<summary>См. пример из нашей текущей доки</summary>

![image](https://user-images.githubusercontent.com/42924400/110370037-3df8e380-805c-11eb-92af-310b27a1dd0c.png)
</details>

> Q: А почему я не могу настраивать нейминг как хочу?
> A: См. топик [Адаптивность нейминга](https://github.com/feature-sliced/wiki/discussions/16)

**Comment by azinit** [comment]
> @ilyaagarkov натолкнул своими вопросами на классную мысль (см. https://github.com/feature-sliced/wiki/discussions/23#discussioncomment-451459 )

> `*AAA` - значит опционально
>
> UPD: `components` ~ `ui`; `store` ~ `models`

## `by types` Раньше (давно-давно) мы делили по типам
- Сначала `components`, `store`, `api`, `lib`
- Потом (внутри каждого) `tasks`, `users`, `posts`

## `by domains` Теперь же мы делим по фичам (или доменам) 
- Сначала `app`, `*processes`, `pages`, `features`, `*entities`, `shared`
- Потом (внутри каждого) `components`, `*store`, `*api`, `*lib`

## Иными словами....
High/middle абстракции можно понимать следующим образом:
- `High-level-abstraction`: абстракция "по функциональности" (методология сначала разбивает по ним)
- `Middle-level-abstraction`: абстракция "по типам" (то как делили раньше, но теперь у нас каждая абстракция делится по типам, и вроде выглядит логично)

**Reply by azinit** [reply]
Опять же это про "торт" и `Separation of Concerns`:

- Мы либо сначала разрезаем **по фичам, а потом по типам** (чаще всего лучше)
- Либо разрезаем сначала **по типам, а потом по фичам** (вложенность ниже, но понимание - стремится к нулю)

**Reply by azinit** [reply]
Отсюда следует простое следствие, что **у нас может быть максимум одно использование каждой абстракции в пути к файлу**

Т.е. можно это представить как такой двойной select:

<details>
<summary>1. <code>BY-DOMAIN</code> Выберите назначение модуля <b>по функциональности</b></summary>

- [ ] `app`
- [ ] `processes`
- [ ] `pages`
- [ ] `features`
- [ ] `entities`
- [ ] `shared`
</details>

<details>
<summary>2. <code>BY-TYPE</code> Выберите назначение модуля <b>по типу</b></summary>

- [ ] `ui` (или `components`)
- [ ] `models` (или `store`)
- [ ] `lib`
- [ ] `api`
</details>


3. Fullname вашего модуля (условно):
- `{BY-DOMAIN}/{BY-TYPE}`

**Reply by azinit** [reply]
<details>
<summary>Примеры "Хорошо"/"Плохо"</summary>

```diff
- features/auth/features/by-google
- shared/**/entities
- ui/feature/auth
- ui/auth
+ features/auth/by-google
+ shared/ui/button
+ entities/skins/ui
```
</details>

**Reply by azinit** [reply]
Хотя бывают и случаи, когда модуль разрастается по логике, и приходится нарезать повторно в том числе и его
> Но не исключено, что это значит только то, что модуль должен сменить свой `BY-DOMAIN` атрибут

Пример:
```sh
# Какой-нибудь сложный компонент, в силу своей сложности - вынужден разрезаться еще раз "по типам"
# Т.к. иначе логика смешается
# Хотя возможно, что в таком случае - компонент уже становиться entity/features/...?
- shared/components
      - smth-difficult-popup
           -  components/
                - source/ ...
                - target/ ...
           -  context/
                - hooks
                - provider
                - index.ts
```

**Reply by azinit** [reply]
Вот тут хорошо описано, поч мы поняли, что "нарезать надо по-другому"

https://ryanlanciaux.com/blog/2017/08/20/a-feature-based-approach-to-react-development/

![image](https://user-images.githubusercontent.com/42924400/111881693-cbc0c100-89c2-11eb-8b89-d2a2f1750f74.png)

**Comment by binjospookie** [comment]
Почему мне не нравится shared:
При первом прочтении появляется желание засунуть туда всё, что может быть переиспользовано. Те же самые `entities`.

Lib -- менее абстрактный неймнинг, кмк, достигающий оригинальную цель (если же правильно её понимаю).

**Reply by azinit** [reply]
Соглашусь, это надо наверное обсудить тоже

Но тут видишь какие нюансы:
1. `lib` - уже является более примитивной абстракцией, поэтому по неймингу не подходит ужи (ты же в lib не хранишь одновременно и `lib/lib` и `lib/ui`

2. `shared` - это вроде как наиболее часто используемое имя (второй наверное `common`) для инфраструктурных **переиспользуемых** вещей. И скорее это просто следующий эволюционный шаг , чтобы ui/lib не на самом верху держать, а в какой-то high-level абстракции-фиче

> Тут опять же другой вопрос - а где гарантии, что люди не будут "сувать туда все что можно" в features / entities и т.д.?

**Reply by binjospookie** [reply]
> ты же в lib не хранишь одновременно и lib/lib и lib/ui

Почему нет? `lib/ui-kit` и `lib/currency` (например).

Здесь я больше опираюсь на то, как мы произносим и думаем:
* ui-kit библиотека
* библиотека для работы с валютами

**Reply by ilyaagarkov** [reply]
> Тут опять же другой вопрос - а где гарантии, что люди не будут "сувать туда все что можно" в features / entities и т.д.

вообще никаких гарантий пока у нас нет  IDE чисто под feature-sliced которая запрещает создавать что либо руками =)

Кстати, а если методологию представить в виде дерева вопросов, когда последовательность ответов приводит тебя к информации о том куда стоит положить тот или иной код?

Скорее всего это будет весьма не просто, но если получится то мы точно можем констатировать факт что мы сами понимаем как это должно работать.

**Reply by azinit** [reply]
> Кстати, а если методологию представить в виде дерева вопросов, когда последовательность ответов приводит тебя к информации о том куда стоит положить тот или иной код?
>
> Скорее всего это будет весьма не просто, но если получится то мы точно можем констатировать факт что мы сами понимаем как это должно работать.

@ilyaagarkov  [welcome](https://github.com/feature-sliced/wiki/discussions/29) ))
> День назад этот ишью заводил как раз

Но пока да - здесь надо прийти к общему пониманию абстракций

**Reply by sergeysova** [reply]
> Кстати, а если методологию представить в виде дерева вопросов, когда последовательность ответов приводит тебя к информации о том куда стоит положить тот или иной код?

Вот это ровно то, к чему я пытаюсь прийти. Когда сортируешь все сущности по сложности внутреннего устройства, обнаруживается и их взаимосвязь и опасность их изменений.

Предлагаю начать составлять такие вопросы. Но кажется, что для этого нужно насобирать примеров

**Reply by sergeysova** [reply]
Проблема shared to lib: В shared лежит конфигурация и api.

```
shared/api
shared/config
shared/lib/currency
shared/lib/username
shared/lib/validations
shared/ui
```

а теперь представим себе всё в куче:

```
lib/api
lib/config
lib/currency
lib/ui
lib/username
lib/validations
```

Всё сразу превращается в свалку.

**Reply by binjospookie** [reply]
Не обязательно:

```
lib/api
lib/config
lib/service/currency
lib/service/username
lib/service/validations
lib/ui
```

**Reply by sergeysova** [reply]
Второй фактор: мы имеем некоторую фрактальную структуру, когда структура shared может повторяться в features и entities:

```
src/
  features/
    <feat-name>/
      api/
      lib/
      ui/
  entities/
    <entity>/
      api/
      lib/
      ui/
  shared/
    api/
    lib/
    ui/
```

То есть мы четко обозначаем **что** мы можем "шарить" в приложении, что достаточно "глобальное" и абстрактное, а что имеет привязку к конкретной сущности или функциональности.

Чтобы не образовывалась свалка в `shared` нужно четко обозначить, что туда можно положить и мб даже задать структуру директорий.

**Reply by sergeysova** [reply]
@binjospookie 

```
lib/api
lib/config
lib/service/currency
lib/service/username
lib/service/validations
lib/ui
```

Это НЕ сервисы, это методы работы с какими-то данными, могут быть весьма разрозненными.
Название service в любом случае вводит в заблуждение.

А любое другое название просто поменяет именование `{shared}/{lib}`, и убьет назначение "разделяемое"

**Reply by sergeysova** [reply]
#### Почему не класть `ui` внутрь `lib`?

Во первых, мы часто используем библиотеки из `lib` при реализации компонентов ui kit, а значит, они уже не могут быть на одном уровне абстракции.
Во вторых, это категорически разные типы библиотек, и при этом имеют очень разное внутреннее устройство.

Тогда как `shared/lib` пытаются быть максимально гранулированными и простыми, `shared/ui` будет устроен совсем по другому и попытается реализовать все необходимые базовые компоненты проекта, даже может использовать AtomicDesign

**Reply by binjospookie** [reply]
Соглашусь с тобой.

**Comment by ilyaagarkov** [comment]
У меня еще появились сомнения по поводу нейминга папки ui. (именно той которая "главная")
Ведь по сути там лежит набор самых базовых компонентов, то есть это не ui, а именно ui-kit

**Reply by azinit** [reply]
А что ты скажешь про компоненты в фиче? Или entity?

Ты ведь их вряд ли называешь `ui-kit`)

**Reply by azinit** [reply]
Я обычно их называю `components`, но ребятам (@sergeysova , @karina-drummer ), вродь больше `ui` нравится

И тут холивар как раз)

**Reply by azinit** [reply]
Тип соль в том что у тебя одна и та же абстракция-тип (ui / lib/ api / model/...) - может встречаться в том или ином виде в абстракциях-фичах (features/pages/shared/...)

**И по-хорошему - оно должно называться везде одинаково**


> Надеюсь правильно твою мысль понял

**Reply by ilyaagarkov** [reply]
точно нет (про ui-kit в фичах) =). Вот там мне больше нравится как раз components.

Тут как я вижу, грань опять проходит по бизнесу, кнопки/инпуты и прочие обезличенные компоненты это ui-kit, а вот компоненты отображающие конкретные сущности или процессы это уже components лежащие где-то внутри того что ты назвал High-level abstractions (за исключением shared)

**Reply by ilyaagarkov** [reply]
> И по-хорошему - оно должно называться везде одинаково

должно ли? Все таки базовые компоненты из ui-kit и компоненты напрямую связанные с бизнес логикой это в корне разные вещи.

**Reply by ilyaagarkov** [reply]
но тут конечно ломается описание про High-level-abstraction и Middle-level-abstraction и что High-level может иметь свои собственные Middle-level .

**Reply by ilyaagarkov** [reply]
Кстати, Middle-level как бы подразумевает что есть еще Low-level. Если в итоге все таки придерживаться данной идет то может лучше High-level и Low-Level?

**Reply by azinit** [reply]
> должно ли? Все таки базовые компоненты из ui-kit и компоненты напрямую связанные с бизнес логикой это в корне разные вещи.

Ну смотри, какая структура, по-твоему покажется понятней тебе и в целом обширному кругу разрабов

1. Стул 1
```ts
- entities/user
   - components/
   - lib/
- shared/
   - components/
   - lib/
```

2. Стул 2 (утрированно, но хорошо суть отображает)
```ts
- entities/user
   - components/
   - libs/
- shared/
   - ui/
   - core-lib/
```

**Reply by azinit** [reply]
> но тут конечно ломается описание про High-level-abstraction и Middle-level-abstraction и что High-level может иметь свои собственные Middle-level .
>
> Кстати, Middle-level как бы подразумевает что есть еще Low-level. Если в итоге все таки придерживаться данной идет то может лучше High-level и Low-Level?

А я там выше написал же, что это не оч определение

В данном случае у нас просто два вида абстракций:
- `ТИПЫ`: components, api, lib, store
- `ФИЧИ`: app, processes, pages, features, entities, shared

И если сравнивать наши проекты с тортом - то мы можем разрезать их двумя способами (см. опять же [выше пост](https://github.com/feature-sliced/wiki/discussions/31#discussioncomment-451612) про это)

- Сначала по типам, потом по фичам (все легаси проекты, стродаем от распыления логики)
- Сначала по фичам, потом по типам (это и есть наш feature-sliced)

**Reply by ilyaagarkov** [reply]
я видимо сел бы где-то мимо стульев, при этом уронив стол =)

```sh
- entities
  - /user
     - components/
     - libs/
- ui-kit/
- libs/
```

В даном случее я так же вижу, что ui-kit и components это вообще о разном. У разработчиков тоже должно что-то щелкать когда они выбирают куда что положить.

C одной стороны, да - это дополнтельное усложнении. С другой,  архитектура это всегда сложно

**Reply by sergeysova** [reply]
Я крайне плохо отношусь к разным названием одного и того же в проекте.

> кнопки/инпуты и прочие обезличенные компоненты это ui-kit, а вот компоненты отображающие конкретные сущности или процессы это уже components

На самом деле нет, разницы нет никакой. Это просто компоненты, которые что-то отображают или делают. Мы уже разделили их по назначению разместив в разных директориях: shared и features/entities.

Нет никакого смысла, на мой взгляд увеличивать и без того высокую ментальную нагрузку добавлением разных имён, потому что это вызовет закономерные вопросы "а почему там components, а тут ui?" или "а если мне надо теперь в feature создать компонент, куда его положить components, ui-ki, ui или core-components?"

Чтобы разработчикам не пришлось отвечать на эти вопросы, мы должны решить эту проблему. И смею заметить мы **её уже решили**: добавив by types и by features.

Если мы договоримся везде именовать `components` я не против, но мне важно, чтобы было консистентно.

**Reply by Lehakos** [reply]
Мне кажется components более обширное понятие, потому как компоненты могут быть не только на отображение чего-то. В Реакте, например, компоненты могут и другие цели выполнять. А вот ui это как-будто просто про отображение

**Comment by akaia-shadowfox** [comment]
Я вернулась в тот лагерь, который за "ui" и никаких "components". Раньше глобальные ui / lib у нас не лежали в shared ( aka common aka core ), а значит было некоторой нагрузкой отличать /ui от /features/sample-name/ui, но теперь у нас есть явный префикс "shared" в первом случае

**Reply by azinit** [reply]
Я думал ты и до этого топила всегда за `ui`))

**Reply by akaia-shadowfox** [reply]
<img width="172" alt="image" src="https://user-images.githubusercontent.com/8805308/111064560-cb598f00-84c5-11eb-9d46-1b802947dcd8.png">
Наткнулась на кейс, который намекает на то, что "ui" лучше чем "components"
Стоит задача сделать из wysiwyg фичу. На данный момент это отдельный пакет, в котором переопределяется и дополняется плагинами third-party редактор ckeditor5. Проблема в том, что в данном примере `view/dialog.js` это не компонент, но имеет отношение к UI, и более того, внутри даже происходит импорт из пакета "@ckeditor/ckeditor5-ui"

**Reply by azinit** [reply]
А поч кста тогда ты не назвала `ui/dialog.js`?

**Reply by azinit** [reply]
Или для тебя `ui` и `view` равноценны +-?

**Reply by akaia-shadowfox** [reply]
> А поч кста тогда ты не назвала `ui/dialog.js`?

не я писала, да и коду кучу лет уже

**Comment by azinit** [comment]
Готе таки попробуем проголосовать

- 👍 За
- 👎 Против
- 👀 Воздерживаюсь

> 1. Также пишите свои предложения, если не находятся из имеющихся
>
> 2. Описание каждого варианта (и его основные сторонники) - есть в основном посте выше
>
> 3. Опрос здесь, а не в тг-чате, т.к. здесь же есть и место для дискуссий, и описание каждой сущности во всей полноте
>    > Пришлось в github, сорь за формат, но пришлось, т.к. в тг не самое подходящее
>
> 4. В опросе расписаны только "конфликтные" варианты, чтобы не засорять лишним
>    > По остальным моментам +- сходимся, но если не согласны с тем - что написано в основном посте - пишите)

**Reply by azinit** [reply]
#### UI-логика и отображение
Предпочитаю `components`

**Reply by azinit** [reply]
#### UI-логика и отображение
Предпочитаю `ui`

**Reply by azinit** [reply]
#### Data-логика, стейт, события абстракции
Предпочитаю `store`

**Reply by azinit** [reply]
#### Data-логика, стейт, события абстракции
Предпочитаю `models`

**Reply by azinit** [reply]
#### Data-логика, стейт, события абстракции
Предпочитаю `controller`

**Reply by azinit** [reply]
#### Data-логика, стейт, события абстракции
Предпочитаю `data`

**Reply by azinit** [reply]
#### Директория для инфраструктурных и переиспользуемых модулей
Предпочитаю `shared`

**Reply by azinit** [reply]
#### Директория для инфраструктурных и переиспользуемых модулей
Предпочитаю `lib`

**Reply by azinit** [reply]
#### Директория для инфраструктурных и переиспользуемых модулей
Предпочитаю `common`

**Reply by azinit** [reply]
#### Директория для инфраструктурных и переиспользуемых модулей
Предпочитаю `ui`/`lib`/`api` выносить на самый вверх, без отдельной директории

**Reply by azinit** [reply]
#### Инициализирующая логика приложения
Предпочитаю `app`

**Reply by azinit** [reply]
#### Инициализирующая логика приложения
Предпочитаю `core`

**Reply by azinit** [reply]
#### Инициализирующая логика приложения
Предпочитаю все хранить в `src/index.tsx` а нужные файлы для инициализации располагать рядом

**Reply by azinit** [reply]
#### Data-логика, стейт, события абстракции

Предпочитаю `state`

**Reply by azinit** [reply]
#### Data-логика, стейт, события абстракции

Бывает, выношу логику в `hooks` / `effects` - и тогда стейт вообще не нужен (как отдельная сущность)

**Reply by akaia-shadowfox** [reply]
#### Директория для инфраструктурных и переиспользуемых модулей

Мои варианты:

- `basic`
- `elements`
- `essentials`
- `ingredients`
- `units`
- `kit`
- `packages`

**Reply by ilyaagarkov** [reply]
**Директория для инфраструктурных и переиспользуемых модулей**

general

**Comment by azinit** [comment]
##  `models` vs `store`
---

UPD: Дискуссия под конец уехала [сюда](https://github.com/feature-sliced/wiki/discussions/68)

---
> @martis-git 
> А можешь еще раз напомнить - слайсы у тебя это отдельная модель (store) в фиче/entity или слайс == конкретная фича/entity?

> @karina-drummer 
> Слайс -- это feature / entity, да
> Модель я так моделью и называю, она лишь часть слайса

> @martis-git 
> Вот почему спрашиваю прост)
> > Привет редаксоидам и redux-toolkit
![image](https://user-images.githubusercontent.com/42924400/110891924-1cba2080-8304-11eb-9b20-49dcc07fa469.png)
>
> Т.е. слайс - это не всегда только про абстракцию, для нас редуксоидов - это значит и кое-что другое 
> > https://redux-toolkit.js.org/api/createSlice

**Reply by azinit** [reply]
> @karina-drummer 
> Понятно что тебя запутало
> Поэтому и имеет смысл использовать общепринятый термин "модель"

> @karina-drummer 
> экшны + эффекты + редьюсер = модель
состояние экземпляра модели в памяти в момент времени = состояние
> Модель это шаблон поведения сущности ( возможные состояния, факторы его изменения ), но не само состояние

**Reply by azinit** [reply]
## А теперь давай по `models` vs `store`

> Мы сейчас говорим про абстракцию `Data-логика, стейт, события абстракции`, можете отдать свой голос [выше](https://github.com/feature-sliced/wiki/discussions/31#discussioncomment-464894) за своего фаворита

### Что я могу сказать про `models`

1. **Слово "модель" мне кажется логичным только в формате парадигм `MV**`**

Да, там действительно Model - это данные (и возможно некая логика), View - отображение
Но не стоит также забывать, что чаще всего большая часть логики еще сосредоточена в Controller/Presenter/... - т.е. 

**Модель чаще всего только про сущности и сами данные, с которыми мы работаем, но не всегда про логику**

2. Отсюда переходим ко второму моему набросу - мне не нравится модель также потому - что у нас уже есть слово `entities` зарезервированное для работы с сущностями/моделями БЛ 

И вот даже у меня, у человека, который погружен в эту методологию и готов идти на компромиссы - все равно диссонанс, когда я вижу такие коллизии имен

3. И ВСЕ. Больше я не понимаю, зачем работу с "данными и логикой" натягивать на одно слово `model`))
    > Давайте тогда и компоненты/ui называть `view`, а `index.ts` файлы или процессы - контроллерами?
    > `MV**` - это хорошо, но как мне кажется в качестве нейминга по методологии - это не оч хороший выбор

### Окей, а как тогда называть работу с данными/эффектами/стейтом/поведением?

Обычно, для нейминга работы с данными я исхожу от того - а в каком виде эта работа с данными представлена?
> Т.е. да, немного завязываюсь на реализацию, но не против и одного единого нейминга

1. `store` - чаще всего юзаю его, т.к. обычно у меня стейт - это часть редаксового store, хотя не вижу ничего плохого и юзать его, где напрямую не используется глобальный store
      > Почему-то когда я вижу в структуре фичи `components` - я понимаю, что это про визуальную составляющую, а когда вижу `store` - сразу понимаю, что здесь сосредоточена логика фичи и работа с ее данными, включая эффекты

2. Пореже юзаю другие варианты:
- `effects.ts` - для определения эффектов, которые "напрямую не меняют store" абстракций, а скорее явно определяют взаимодействие между сущностями 
    > "если произошло feature1.action1 - то сделай feature2.action2" и т.д.

- `context` - обычно, когда явно используется React.Context, хотя с тем же успехом мог назвать его и `store`
    >  Обычно когда контекст большой - там все равно есть отдельно работа со стейтом - включая те же экшены, редьюсер - и иногда у меня внутри папки `**/context` встречается модуль `store` - как мини абстракция, хотя возможно можно и проще


---

## Что по итогу?

- Понятно, что не очень хочется завязываться на реализации, при использовании имени
- Таже хочется подобрать для работы со стейтом такое имя, которое бы подходило в большинстве случаев и не конфликтовало с другими (привет `models` и `entities` - из-за чего у нас и произошел ренейминг)

**Reply by akaia-shadowfox** [reply]
>Да, там действительно Model - это данные

Полное название данной концепции -- "модель данных", то есть это не сами данные, а шаблон их поведения, макет

**Reply by akaia-shadowfox** [reply]
>Давайте тогда и компоненты/ui называть view

Во времена MVC ещё толком не существовало комплексной концепции ui-компонентов, как правило views там представляли собой pages, которые иногда собирались из snippets / templates / etc., которые по сути и были чем-то вроде компонентов. Кстати, я уже когда-то предлагала использовать "views" вместо "pages", чтобы сделать название "экранов" универсальным для интерфейсов на любых платформах ( за пределами web то же самое что и page может называться как screen / activity / etc. )

**Reply by ilyaagarkov** [reply]
Проблемы с `store`- название никоим образом не отображает содержимое.

Для пользователей `redux`(точнее все синглстороровых решений) существует только один store, а у нас выглядит так будто в каждой папке свой. Хотя на деле мы описываем только часть редьюсеров и экшенов

Для пользователей `effector` (или остальных мультисторовых решений) - store так же не подходит, так как скорее всего там будет лежать много сторов

`model` по мне куда более удачное название, все сразу понимают о чем речь, речь о данных. Не сказал бы что на фронте слово model ассоциируется с кусочком из MVC, model давно уже живет своей жизнью.

Как альтернатива `model`,  разве что `state` приходит в голову

**Reply by dpr-dev** [reply]
> ### Что я могу сказать про `models`
> 1. **Слово "модель" мне кажется логичным только в формате парадигм `MV**`**
> 
> Да, там действительно Model - это данные (и возможно некая логика), View - отображение
> Но не стоит также забывать, что чаще всего большая часть логики еще сосредоточена в Controller/Presenter/... - т.е.
> 
> **Модель чаще всего только про сущности и сами данные, с которыми мы работаем, но не всегда про логику**

совершенно не так. чаще всего в VM/** паттернах модель - вся бизнесовая логика приложения. Она может быть как анемичной, так и богатой. В случае анемичной модели у нас действительно появляются некие "сущности", к-ые по факту просто служат контейнером данных, а вся логика сосредоточена в сервисах, менеджерах, интеракторах, доменных сервисах и тд / тп, как угодно называйте. Но это всё ещё модель 

презентеры / контоллеры / вью-модели - как раз то, что непосредственно к БЛ отношения не имеет. это фасад, который говорит, как использовать существующую БЛ, связывает её с инфраструктурным кодом, адаптирует данные для вью и тд

**Reply by dpr-dev** [reply]
> 1. И ВСЕ. Больше я не понимаю, зачем работу с "данными и логикой" натягивать на одно слово `model`)

а что тогда есть бизнес-логика ? В вашей интерпретации, как дифференцировать работу с данными на инфраструктурный код (круд - тоже инфраструктура), бизнес-логику, UI логику и вспомогательные хелперы какие-то, к-ые идут как сквозные зависимости

**Reply by azinit** [reply]
@dpr-dev спасибо, что расписали

Да, действительно сам начинаю потихоньку проникаться больше версией `model` для обозначения бизнес-логики

Но все еще кажется - что как будто ни `model`, ни `store` - не подходят до конца...
*https://github.com/feature-sliced/wiki/discussions/31#discussioncomment-477800*

Надо подумать еще однозначно, просто не хочется смешивать доменные абстракции (entities) с типовыми (model)
*(Наверняка будет путанница, а хотелось бы простоты и очевидности для разработчиков)*

**Reply by azinit** [reply]
- Есть скорее ощущение что фраза "model == бизнес-логика" в контексте `MVC` - еще кажется логичной вполне
- Но когда мы погружаемся в дремучую разработку, где есть модели с API и другие термины (не только с MVC) - такое решение кажется уже не очень очевидным
     - Потому, например, слово `view` на фронте приходится использовать крайне редко

**Reply by binjospookie** [reply]
А если не model и не store, а struct?

Просто набрасываю

**Reply by azinit** [reply]
> В вашей интерпретации, как дифференцировать работу с данными на инфраструктурный код (круд - тоже инфраструктура), бизнес-логику, UI логику и вспомогательные хелперы какие-то, к-ые идут как сквозные зависимости

Мы пока плюс-минус единогласно понимаем, что каждая абстракция-домен (app, pages, features, ...) **может в себе содержать**:
- `api` - логика **запросов к API** и возможно кеширования данных
- `ui` - логика **отображения**, обычно отображает и подписывается на изменение данных
- `lib` - логика **инфраструктурная**, необходимая для построения БЛ в других аспектах
- `models` - логика **бизнеса**,  реализующая функционал через стейт, эффекты, реактивность и т.д.

> Думаю сейчас все мы в core-team понимаем **какие абстракции нужны в методологии и как их "разрезать"**
*Но по неймингу...* Все еще пытаемся подобрать удачные названия, чтоб возникало меньше таких обширных тредов и вопросов 
(я до сих пор не могу сказать что "за models", но при прочих равных скорее "не так уж и плохо")

**Reply by azinit** [reply]
> Просто набрасываю

@binjospookie так победим)

Я вот тож хз уже какое название взять для модулей с "бизнес-логикой" (aka data-layer?)
https://github.com/feature-sliced/wiki/discussions/31#discussioncomment-477800

**Reply by azinit** [reply]
## Снова сцепились

> @martis-git:
> Если вот начистоту - то меня почти убедили касаемо ui (c коллегами тоже обсуждали, но там это не так критично)
> И чуть поменьше убедили касаемо models -  но там скорее "понимаю почему так, но хотелось бы получше название"
> 
> Мне бы наверное даже и зашло models, если бы только у нас еще своих моделей на фронте с БЛ не было


> @karina-drummer 
> Модель это контекстонезависимая абстракция, которая одинаково используется как в бл, так и в чисто логике приложения
> https://ru.wikipedia.org/wiki/Модель_данных


> @martis-git 
> 1. модель == аспект, отвечающий за логику в системе (MVC)
> 2. модель == ~сущность из бизнес-логики системы (БЛ)
>
> вот эта коллизия меня наверн больше всего и убивает

> @karina-drummer 
> По идее тут нет коллизии, второе обеспечивается первым

> @martis-git 
> Это как назвать папку "logic"
> Очевидно, что будут абстракции, которые порождаются этим понятием
> И из-за этого теряются наши относительные ["явность", "очевидность", "простота"](https://github.com/feature-sliced/wiki/blob/master/about/architecture.md#explicit-%D0%BF%D0%BE%D0%BD%D1%8F%D1%82%D0%BD%D0%BE%D1%81%D1%82%D1%8C%D1%8F%D0%B2%D0%BD%D0%BE%D1%81%D1%82%D1%8C)

**Comment by azinit** [comment]
## `shared` vs `common`

**Reply by azinit** [reply]
**Недавно пришлось поработать плотно с UNIX системами, и вспомнил - что даже там понятие `shared` используется гораздо чаще, когда мы говорим про "общие переиспользуемые ресурсы"**

> Само собой, речь даже не только про UNIX, а в целом про администрирование ресурсов и доступа

> https://refspecs.linuxfoundation.org/FHS_2.3/fhs-2.3.pdf

**Comment by azinit** [comment]
# Нужно ли жестко фиксировать наименование для работы с данными?

> `models`, `store`, `state`, `data`, ...

**Reply by azinit** [reply]
Пока по нашим результатам опросов - у всех оч разное мнение, как именовать модуль для работы с данными-логикой-эффектами

**Reply by azinit** [reply]
Хотелось бы, чтобы была одна папка `{DATA_LAYER_NAME}` **со всей логикой домена**, а в ней уже нужные абстракции, привязанные к реализации стейт-менеджмента и к тому- как это видит разработчик/архитектор:

```sh
- feature/auth-form
    - ui/
    - {DATA_LAYER_NAME}/
         -  index.ts
         -  *thunks
         -  *effects
         -  *slice
         -  *actions
         -  *controllers
         -  *epics
         -  *context
         -  ...
```

**Reply by sergeysova** [reply]
Так пусть она называется `model`?
```
- feature/auth-form
    - ui/
    - model/
         -  index.ts
         -  *thunks
         -  *effects
         -  *slice
         -  *actions
         -  *controllers
         -  *epics
         -  *context
         -  ...
```

**Reply by azinit** [reply]
@sergeysova вот почему)

![image](https://user-images.githubusercontent.com/42924400/111298692-19a29580-8660-11eb-8add-da906d0b3aab.png)
